### DISCUSSION TITLE
How should I distribute tagfiles?

### DISCUSSION THREAD

**Discussion by meator** [doxygen/doxygen#11295] [Q&A]
Hey. I am working on a tree of libraries, all of which use Doxygen for managing documentation. I would like to link between these gracefully as described in https://www.doxygen.nl/manual/external.html. This page is informative, but it lacks the answer to (in my opinion) the most important question: where should I get the tagfile from?

I am not asking how to generate tagfiles. That part of the documentation is clear.

Let's say that I have two libraries, `parent` and `child`. `child` depends on `parent`. I would like to reference things from `parent` in `child`. `child` therefore needs `parent`'s tagfile. From where should `child` get the tagfile? I see these options:

1. `child` would vendor `parent`'s tagfile

   I would have to manually make sure that it is up to date. This is error prone but simple to implement.
2. tagfile would be downloaded by the build process

   It could be provided alongside online documentation or as a release artifact of `parent`.

   This approach feels unnatural for me. Ideally, the build system should not require an internet connection once all source files are downloaded (from release archives etc.). I do not count `parent`'s tagfile as a source file (unless it is vendored, in which case it would fall into 1.).
3. `parent` would distribute its tagfile (`parent`'s install target would install the tagfile in addition to the library, header files etc.)

   This would ideally be part of a development package in distros (usually suffixed by `-dev`, `-devel` or similar). It could be put into `/usr/share/doc/` or into similar locations.

   I think that this is a pretty nice approach, but it is fairly nonstandard. `/usr/share/doc` is generally used for user facing documentation, so using it might not be appropriate for a tag file. What other directory should I choose?

How do other projects handle this? What is the best solution out of these? Did I forget/overlook something?

EDIT:
To clarify, `parent` may either be compiled with `child` or it might be provided externally. I use Meson's Wrap dependency mechanism, which will look for `parent` (in this case) in standard system directories with the help of `pkg-config` and if it can't find it there, it will download `parent`'s source and build it as part of `child`'s build process. This is handled transparently by Meson. The generated documentation should be **identical** regardless of whether `parent` is system installed or not, using tagfiles only when `parent` is built as part of `child` is not tolerable for me.

**Comment by albert-github** [comment]
Small comment:
It is all a bit in how everything is handled and I would at least say keep the tag file and the HTML output corresponding with it close together. I also think that the tag file is a source file (for the documentation it is).

Maybe some inspiration can also be found in the way Qt is handling it and also see https://forum.qt.io/topic/143679/list-of-tags-files-for-doxygen-for-qt5-and-qt6

**Reply by meator** [reply]
Thanks for your response! My current solution (which may still be subject to change) looks as follows:

If my library is building documentation, it will install it into `/usr/share/doc/<mylibname>` (the `/usr/share/doc/` part is configurable using variable `docdir`). It will also install tags into `/usr/share/doc/<mylibname>/doxygen-tagfile.xml`.

The `/usr/share/doc/<mylibname>` directory is marked as `doc` (meant for documentation) and the `/usr/share/doc/<mylibname>/doxygen-tagfile.xml` file specifically is marked as `devel` (development files). Meson categorizes the files it installs.

When building documentation, a custom variable is added into my library's pkg-config file. It is the `docdir` mentioned above. I have also implemented a way to handle everything when my library is downloaded and compiled by Meson as a dependency.

I am still open to replies to this thread. Installing the tagfile alongside library files is a fairly clean solution, but it is pretty nonstandard.
